### PR TITLE
ast+compile: Support entrypoint annotations

### PR DESCRIPTION
**Pull request by philipaconrad**
This PR implements basic support for entrypoint rule/package annotations in Rego. This feature means that optimized bundle builds, WASM, and Plan target compilation all no longer require a CLI argument for the entrypoint, so long as at least one rule was annotated as an entrypoint in the policies provided.

Fixes: #3459

-----

Remaining Work:

 - [x] New annotation struct field + helpers/validation.
 - [x] Entrypoint extraction in `compile`.
 - [x] Re-plumb CLI commands to use the extraction utility, or extract the entrypoints during parsing.
 - [x] Update relevant tests + add new testcases for CLI commands.

### CODE REVIEW

File: ast/annotations.go [hunk 1]
```diff
@@ -301,6 +309,8 @@ func (a *Annotations) Copy(node Node) *Annotations {
 		cpy.Schemas[i] = a.Schemas[i].Copy()
 	}
 
+	cpy.Entrypoint = a.Entrypoint
```
Comment (srenatus) [inline ast/annotations.go:312]: Isn't that covered by line 292, `cpy := *a`?

File: compile/compile.go [hunk 2]
```diff
@@ -210,17 +219,72 @@ func (c *Compiler) WithMetadata(metadata *map[string]interface{}) *Compiler {
 	return c
 }
 
+func addEntrypointsFromAnnotations(c *Compiler, ar []*ast.AnnotationsRef) error {
+	for _, ref := range ar {
+		scope := ref.Annotations.Scope
+		entrypoint := ast.EmptyRef()
```
Comment (srenatus) [inline compile/compile.go:225]: Why do we need this? For some reason `var entrypoint ast.Ref` seems to be insufficient..? 🤔

File: compile/compile.go [hunk 3]
```diff
@@ -302,21 +366,27 @@ func (c *Compiler) init() error {
...
+func (c *Compiler) checkNumEntrypoints(afterBundleLoad bool) error {
+	if !c.useRegoAnnotationEntrypoints || afterBundleLoad {
+		if c.optimizationLevel > 0 && len(c.entrypointrefs) == 0 {
+			return errors.New("bundle optimizations require at least one entrypoint")
 		}
-	case TargetPlan:
-		if len(c.entrypointrefs) == 0 {
-			return errors.New("plan compilation requires at least one entrypoint")
+
+		switch c.target {
```
Comment (srenatus) [inline compile/compile.go:379]: Can we make that `if c.target != TargetRego`? We don't need the `wasm/plan` in the error message, I think. Or we replace that with `c.target`, whatever it is.

File: compile/compile.go [hunk 4]
```diff
@@ -210,17 +219,72 @@ func (c *Compiler) WithMetadata(metadata *map[string]interface{}) *Compiler {
...
+
 // Build compiles and links the input files and outputs a bundle to the writer.
 func (c *Compiler) Build(ctx context.Context) error {
 
 	if err := c.init(); err != nil {
 		return err
 	}
 
+	// Fail early if not using Rego annotation entrypoints.
+	if err := c.checkNumEntrypoints(false); err != nil {
```
Comment (srenatus) [inline compile/compile.go:256]: I'm a little confused by the interplay between the `bool` argument here, and c.useRegoAnnotationEntrypoints` -- I suspect that it could be rearrange a bit to be more straightforward, but I trust that you've tried that already 🙃

**Review by srenatus — COMMENTED** [review]
looks good so far!

**Comment by anderseknert** [comment]
I'll come back here when there are tests :) But I'm curious what the intention is when there are multiple sources of entrypoints. Are they meant to be additive? So I can specify one in an annotation, and another one via the CLI?

**Comment by philipaconrad** [comment]
> ... when there are multiple sources of entrypoints. Are they meant to be additive? So I can specify one in an annotation, and another one via the CLI?

Correct. That's *exactly* the intent. :smile:

I looked at the plumbing involved, and realized that as long as I'm careful to do the "do we have any valid entrypoints available?" check before compiling stuff, we can have more than one source of entrypoints without issue.

The only side-effecting behavior I'm worried about is that now modules *always* parse annotations (they apparently didn't before during bundle builds!), since this feature is enabled by default currently. If there's any real risks around parsing annotations that I'm not aware of, I could feature-gate it behind a CLI flag, but I'd love for this to become standard behavior for bundle builds going forward.

File: compile/compile.go (hunk 3 above)
Comment (philipaconrad) [reply compile/compile.go:379]: This was just a straight port of the logic that was already there--which generated per-target error messages. I like the idea of parametrizing the error message though. I'll roll that in with my next batch of changes.

**Comment by philipaconrad** [comment]
Commits have been squashed.

Tests have been provided at 2x levels: `compile` package tests, and CLI tests for `build` and `eval` commands, to ensure that they correctly pick up entrypoint annotations.

File: compile/compile_test.go [hunk 5]
```diff
@@ -821,6 +821,201 @@ func TestCompilerPlanTargetUnmatchedEntrypoints(t *testing.T) {
...
+				// Ensure those entrypoints match the ones we expect.
+				for _, entrypoint := range compiler.entrypoints {
+					if _, found := tc.wantEntrypoints[entrypoint]; !found {
+						t.Fatalf("Unexpected entrypoint '%s'", entrypoint)
+					}
+				}
+			})
+		})
+	}
+}
```
Comment (srenatus) [inline compile/compile_test.go:1017]: Let's have a ref head rule, too! 😃

**Review by srenatus — DISMISSED** [review]
LGTM just one nitpick 👍

**Comment by philipaconrad** [comment]
For some reason, only this handful of E2E WASM tests from `opa/test/e2e/wasm/authz/authz_bench_integration_test.go` are failing. They have nearly identical siblings in `opa/test/e2e/authz/authz_bench_integration_test.go` that run just fine.

As best I can figure, something may be weirdly out-of-sorts around the API calls the WASM tests use to set up their environment. :man_shrugging: 

```
$ go test -count=1 -tags=opa_wasm,slow -bench=. github.com/open-policy-agent/opa/test/e2e/wasm/authz
--- FAIL: BenchmarkRESTAuthzForbidAuthn
    authz_bench_integration_test.go:163: unexpected response status: 500 Internal Server Error
--- FAIL: BenchmarkRESTAuthzForbidPath
    authz_bench_integration_test.go:163: unexpected response status: 500 Internal Server Error
--- FAIL: BenchmarkRESTAuthzForbidMethod
    authz_bench_integration_test.go:163: unexpected response status: 500 Internal Server Error
--- FAIL: BenchmarkRESTAuthzAllow10Paths
    authz_bench_integration_test.go:163: unexpected response status: 500 Internal Server Error
--- FAIL: BenchmarkRESTAuthzAllow100Paths
    authz_bench_integration_test.go:163: unexpected response status: 500 Internal Server Error
FAIL
exit status 1
FAIL	github.com/open-policy-agent/opa/test/e2e/wasm/authz	1.274s
FAIL
```

**Comment by philipaconrad** [comment]
Figured it out! I was feeding dotted paths into the planner in a spot where slashed paths were required. The affected line was a holdover from earlier experimentation, before I was generating slash-paths correctly with `.Ptr()`.

File: compile/compile_test.go (hunk 5 above)
Comment (philipaconrad) [reply compile/compile_test.go:1017]: This was a great suggestion, because as I had feared, I wasn't using `ref.Head.Ref()` to extract the refs for entrypoints, and was instead relying on the older `ref.Head.Name` style that can fail on newer-style ref head rules.

Comment (philipaconrad) [reply compile/compile_test.go:1064]: Rolled in 2x new tests for ref heads. @srenatus if you can think of more clever cases I should try, please LMK, and I'll roll them in to another round of changes. :smile:

File: cmd/build_test.go [hunk 6]
```diff
@@ -292,6 +292,72 @@ func TestBuildPlanWithPruneUnused(t *testing.T) {
 	})
 }
 
+func TestBuildPlanWithRegoEntrypointAnnotations(t *testing.T) {
+
+	tests := []struct {
+		note       string
+		files      map[string]string
+		shouldFail bool
```
Comment (srenatus) [inline cmd/build_test.go:300]: I think it could be worth expanding the failure case, maybe? i.e.

```suggestion
		shouldFail error // use nil if none expected
```

File: cmd/eval_test.go [hunk 7]
```diff
@@ -676,6 +676,38 @@ func removeBuiltin(builtins []*ast.Builtin, name string) []*ast.Builtin {
...
+	files := map[string]string{
+		"test.rego": `
+package test
+default p = false
+# METADATA
+# entrypoint: true
+p { q }
+q { input.x = data.foo }`,
+		"data.json": `
+			{"foo": 1}`,
```
Comment (srenatus) [inline cmd/eval_test.go:692]: [nit] Funky indentation for a verbatim string. But it doesn't matter, of course.

File: cmd/eval_test.go [hunk 8]
```diff
@@ -676,6 +676,38 @@ func removeBuiltin(builtins []*ast.Builtin, name string) []*ast.Builtin {
...
+	}
+
+	test.WithTempFS(files, func(path string) {
+		params := newEvalCommandParams()
+		params.bundlePaths = repeatedStringFlag{
```
Comment (srenatus) [inline cmd/eval_test.go:697]: Do we not have [a helper ](https://github.com/philipaconrad/opa/blob/7faaab38fe70960cdc6cb0c958d339dbb272ba1a/cmd/eval.go#L706) for this? I think we can use `params.bundlePaths.Set(path)` here.

**Review by srenatus — DISMISSED** [review]
LGTM, nitpicks only.

I think some docs are useful, otherwise this fine UX improvement might go unnoticed.

File: cmd/eval_test.go (hunk 8 above)
Comment (philipaconrad) [reply cmd/eval_test.go:697]: I was entirely unaware that there was a helper for that, but I'll check it out. Thank you for linking it!

File: cmd/eval_test.go (hunk 7 above)
Comment (philipaconrad) [reply cmd/eval_test.go:692]: During development, I had issues around metadata comments not being allowed to have spaces in front of them.
Example from `cmd/build_test` after putting "proper" indentation around the verbatim string:
```
Running tool: /usr/local/go/bin/go test -timeout 30s -tags opa_wasm,slow -run ^TestBuildPlanWithRegoEntrypointAnnotations$ github.com/open-policy-agent/opa/cmd

--- FAIL: TestBuildPlanWithRegoEntrypointAnnotations (0.00s)
    --- FAIL: TestBuildPlanWithRegoEntrypointAnnotations/annotated_entrypoint (0.00s)
        /dev-folder/opa/cmd/build_test.go:348: load error: 1 error occurred during loading: /tmp/opa_test3450830366/test.rego:2: rego_parse_error: expected METADATA block, found whitespace
            	# METADATA
            	^
FAIL
FAIL	github.com/open-policy-agent/opa/cmd	0.012s
FAIL
```

Comment (philipaconrad) [reply cmd/eval_test.go:692]: The error is caused by [this check](https://github.com/open-policy-agent/opa/blob/482769dd12a1f346780ff0a26ccf2c50480f9547/ast/parser.go#L2146-L2148) in `parser.go`. I'm not sure why we can't have whitespace before a metadata annotation comment, but it seems unnecessarily limiting.  :shrug:

Comment (srenatus) [reply cmd/eval_test.go:692]: I've just been referring to the data.json. but you've got a point there, I suppose.

Comment (philipaconrad) [reply cmd/eval_test.go:692]: Ah, okay. The `data.json` indentation is easy enough to fix, and I'll roll that into my next batch of changes. Addressing the "whitespace breakage" issue is definitely material for a different PR if it's not intentional behavior.

File: cmd/eval_test.go (hunk 8 above)
Comment (philipaconrad) [reply cmd/eval_test.go]: Rolled in a refactoring to use the helper methods consistently in that test file. :+1:

File: cmd/build_test.go (hunk 6 above)
Comment (philipaconrad) [reply cmd/build_test.go]: Rolled in the type change into my latest changes. Thanks for the suggestion! :slightly_smiling_face:

**Comment by philipaconrad** [comment]
Commits have been pre-squashed to make merging easier later.

File: ast/annotations.go [hunk 9]
```diff
@@ -324,6 +332,10 @@ func (a *Annotations) toObject() (*Object, *Error) {
 		obj.Insert(StringTerm("title"), StringTerm(a.Title))
 	}
 
+	if a.Entrypoint {
+		obj.Insert(StringTerm("entrypoint"), BooleanTerm(a.Entrypoint))
```
Comment (srenatus) [inline ast/annotations.go:336]: [nit] Since we only include this if it's true, we could also use 
```suggestion
		obj.Insert(StringTerm("entrypoint"), BooleanTerm(true))
```

File: cmd/build.go [hunk 10]
```diff
@@ -284,6 +287,7 @@ func dobuild(params buildParams, args []string) error {
 		WithOptimizationLevel(params.optimizationLevel).
 		WithOutput(buf).
 		WithEntrypoints(params.entrypoints.v...).
+		WithRegoAnnotationEntrypoints(true).
```
Comment (srenatus) [inline cmd/build.go:290]: :+1: This is a decision I'd agree with: by default, entrypoint metadata will be effective. There's no way to turn it off, but you could just not use it. Since it's not been an allowed keyword before (for metadata), no existing policy should be affected by this.

File: cmd/eval_test.go [hunk 11]
```diff
@@ -730,12 +758,11 @@ func TestEvalWithBundleDuplicateFileNames(t *testing.T) {
 	test.WithTempFS(files, func(path string) {
 
 		params := newEvalCommandParams()
-		params.bundlePaths = repeatedStringFlag{
-			v: []string{
-				filepath.Join(path, "a"),
-				filepath.Join(path, "b"),
-			},
-			isSet: true,
```
Comment (srenatus) [inline cmd/eval_test.go:738]: Thanks for the cleanup 👏

File: compile/compile.go [hunk 12]
```diff
@@ -210,17 +219,83 @@ func (c *Compiler) WithMetadata(metadata *map[string]interface{}) *Compiler {
...
+			case "rule":
+				if r := ref.GetRule(); r != nil {
+					entrypoint = r.Ref()
+				}
+			default:
+				continue // Wrong scope type. Bail out early.
+			}
+
+			// Get a slash-based path, as with a CLI-provided entrypoint.
+			slashPath, err := entrypoint.Ptr()
```
Comment (srenatus) [inline compile/compile.go:243]: Things I'd test here are edge cases involving numbers in refs. Currently, only this should be possible:

```rego
package foo
#METADATA
entrypoint: true
bar.baz[1.0] = true
```
does it become the correct entrypoint? I'm wondering because we've had `array/0/foo` issues with the storage API before, and this is _somewhat_ related... but it's an edge case for sure.

**Review by srenatus — APPROVED** [review]
Thanks! LGTM.

File: ast/annotations.go (hunk 9 above)
Comment (philipaconrad) [reply ast/annotations.go:336]: Rolled this suggestion into my latest set of changes. :+1:

File: compile/compile.go (hunk 12 above)
Comment (philipaconrad) [reply compile/compile.go:243]: It errors currently, which is probably _not_ what we want. :sweat_smile:

Comment (philipaconrad) [reply compile/compile.go]: Okay, after borrowing the path escaping logic from `storage.path`, the above `bar.baz[1.0]` example would generate an entrypoint of `foo/bar/baz/1.0`

For longer paths that want a fully-variable part in the middle, they'll have to convert the variable to a string first. From what I can tell, the logic I borrowed from `storage` is fine, so long as only the last path segment in the ref is variable.

File: compile/compile_test.go [hunk 13]
```diff
@@ -821,6 +821,292 @@ func TestCompilerPlanTargetUnmatchedEntrypoints(t *testing.T) {
...
+q[2]
+q[3]
+				`,
+			},
+			wantEntrypoints: map[string]struct{}{
+				"test/a/b/1.0": {},
+			},
+		},
+		{
+			note:        "string pathj with brackets annotation",
```
Comment (srenatus) [inline compile/compile_test.go:1056]: typo
```suggestion
			note:        "string path with brackets annotation",
```

**Review by srenatus — APPROVED** [review]
LGTM 🥳
